File: 698_partition_to_K_equal_sum_subsets.java
class Solution {
    public boolean canPartitionKSubsets(int[] nums, int k) {
        if (nums == null || nums.length == 0 || k == 0) return false;
        Arrays.sort(nums);
        int sum = 0;
        for (int n: nums) sum += n;
        if (sum % k != 0 || sum < k) return false;
        sum /= k;
        return partition(sum, new int[k], nums, nums.length - 1);
    }

    private boolean partition(int sum, int[] accumulates, int[] nums, int idx){
        if (idx == -1) {
            // examine each accumulation equals to sum or not
            for (int s: accumulates) if (s != sum) return false;
            return true;
        }
        int num = nums[idx];
        for (int i = 0; i < accumulates.length; i++) {
            if (accumulates[i] + num <= sum) {
                accumulates[i] += num;
                if (partition(sum, accumulates, nums, idx - 1)) return true;
                accumulates[i] -= num;
            }
        }
        return false;
    }
}

File: python/698_partition_to_k_equal_sum_subsets.py
class Solution:
    # NP-hard question...should be well-prepared
    #
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        def backtrace(nums: List[int], groups: List[int], target: int, index: int):
            if index == -1:
                for s in groups:
                    if s != target: return False
                return True

            num = nums[index]
            for i in range(len(groups)):
                if groups[i] + num <= target:
                    groups[i] += num
                    if backtrace(nums, groups, target, index - 1):
                        return True
                    groups[i] -= num
            return False

        nums.sort()
        accum = sum(nums)
        if accum % k != 0 or k > accum:
            return False
        target = accum / k
        return backtrace(nums, [0 for _ in range(k)], target, len(nums) - 1)




